client/client.go

```diff
@@ -338,18 +429,107 @@ func (c *Client) getLocalMeta() error {
 		c.loadTargets(targets.Targets)
 	}
 
-	if timestampJSON, ok := meta["timestamp.json"]; ok {
-		timestamp := &data.Timestamp{}
-		if err := c.db.UnmarshalTrusted(timestampJSON, timestamp, "timestamp"); err != nil {
+	c.localMeta = meta
+	return nil
+}
+
+// loadAndVerifyLocalRootMeta decodes and verifies root metadata from
+// local storage and loads the top-level keys. This method first clears
+// the DB for top-level keys and then loads the new keys.
+func (c *Client) loadAndVerifyLocalRootMeta() error {
+	meta, err := c.local.GetMeta()
+	if err != nil {
+		return err
+	}
+	rootJSON, ok := meta["root.json"]
+	if !ok {
+		return ErrNoRootKeys
+	}
+	// unmarshal root.json without verifying as we need the root
+	// keys first
+	s := &data.Signed{}
+	if err := json.Unmarshal(rootJSON, s); err != nil {
+		return err
+	}
+	root := &data.Root{}
+	if err := json.Unmarshal(s.Signed, root); err != nil {
+		return err
+	}
+	ndb := verify.NewDB()
+	for id, k := range root.Keys {
+		if err := ndb.AddKey(id, k); err != nil {
+			// TUF is considering in TAP-12 removing the
+			// requirement that the keyid hash algorithm be derived
+			// from the public key. So to be forwards compatible,
+			// we ignore `ErrWrongID` errors.
+			//
+			// TAP-12: https://github.com/theupdateframework/taps/blob/master/tap12.md
+			if _, ok := err.(verify.ErrWrongID); !ok {
+				return err
+			}
+		}
+	}
+	for name, role := range root.Roles {
+		if err := ndb.AddRole(name, role); err != nil {
 			return err
 		}
-		c.timestampVer = timestamp.Version
 	}
-
-	c.localMeta = meta
+	// Any trusted local root metadata version must be greater than 0.
```

Comment:
The client.rootVer variable needs to be updated with the trusted meta root version. This needs to be done before the Verify err in case that meta is expired.

Else if we start with a root of version 5, the download algorithm will try to load root 1,2,3..+. This will likely fail as root5 keys do not sign root1 after a rotation

Reply:
Nice catch. Updated the method to set the rootVersion -- although we are updating in the updateRoots as well.
But this is the right thing to do.